// @ciscospark/react-component-spaces-list/src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {AutoSizer, List} from 'react-virtualized';
import 'react-virtualized/styles.css';

import SpaceItem from '@ciscospark/react-component-space-item';

const propTypes = {
  activeSpaceId: PropTypes.string,
  hasCalling: PropTypes.bool,
  onCallClick: PropTypes.func,
  onClick: PropTypes.func,
  spaces: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.shape({
        avatarUrl: PropTypes.string,
        activityText: PropTypes.oneOfType([
          PropTypes.string,
          PropTypes.element
        ]),
        callStartTime: PropTypes.number,
        id: PropTypes.string,
        isDecrypting: PropTypes.bool,
        isUnread: PropTypes.bool,
        lastActivityTime: PropTypes.string,
        name: PropTypes.string,
        teamColor: PropTypes.string,
        teamName: PropTypes.string,
        type: PropTypes.string
      })
    ),
    // Can accept an object that can iterated with `forEach`
    PropTypes.object
  ])
};

const defaultProps = {
  activeSpaceId: '',
  hasCalling: false,
  onCallClick: () => {},
  onClick: () => {},
  spaces: []
};

export default function SpacesList({
  activeSpaceId,
  hasCalling,
  onCallClick,
  onClick,
  spaces
}) {
  function rowRenderer(options) {
    const {
      key,
      index,
      style
    } = options;
    const space = spaces[index];

    return (
      <div key={key} style={style}>
        <SpaceItem
          active={space.id === activeSpaceId}
          hasCalling={hasCalling}
          key={key}
          onCallClick={onCallClick}
          onClick={onClick}
          {...space}
        />
      </div>
    );
  }

  return (
    <div className={classNames('ciscospark-spaces-list')} style={{height: '100%', width: '100%'}}>
      <AutoSizer>
        {({height, width}) => (
          <List
            width={width}
            height={height}
            rowCount={spaces.length}
            rowHeight={70}
            rowRenderer={rowRenderer}
          />
        )}
      </AutoSizer>
    </div>
  );
}

SpacesList.propTypes = propTypes;
SpacesList.defaultProps = defaultProps;
